README.md
[![pub.dev Downloads](https://img.shields.io/pub/v/dart_encrypter)](https://pub.dev/packages/dart_encrypter)
[![License](https://img.shields.io/github/license/berkbb/dart_encrypter?color=important)](https://pub.dev/packages/dart_encrypter)


# Encrypter for Dart & Flutter

* Encrypt / Decrypt the String with given Key and IV.

## Features

* Encrypt / Decrypt the String with given Key and IV. It includes built-in password generator also. It uses colored box for outputs if you want.


## Usage
 
1. Add this to your package's pubspec.yaml file:

```dart
dart_encrypter: ^1.0.4
```

2. Save the pubspec.yaml file. Or alternatively you can use this code snippet on console:

```dart
pub get
```

with Flutter:

```dart
flutter pub get
```

3. Import it from code file:

```dart
import 'package:dart_encrypter/dart_encrypter.dart';
```






CHANGELOG.md
## [1.0.4] - 2022-03-15

- Code formatting.

## [1.0.3] - 2022-03-11

- Password generator loop fixed.

## [1.0.2] - 2022-03-11

- Password generator fixed.

## [1.0.1] - 2022-03-10

- generatePasswordwithPrefix() added.


## [1.0.0] - 2022-03-10

- Initial version.


test/dart_encrypter_test.dart
import 'package:dart_encrypter/dart_encrypter.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final _key =
        Security.generatePassword(true, true, true, true, 32); //32 chars
    final iv =
        Security.generatePassword(true, true, true, true, 16); // 16 chars.
    String? item;
    setUp(() {
      // Additional setup goes here.
      item = "Hello my dear !".encryptMyData(_key!, iv!); //Example text.
    });

    test('Encryption test', () {
      expect(item, isNotNull);
      expect(item?.decryptMyData(_key!, iv!), isNotNull);
    });
    test('Create password test', () {
      expect(Security.generatePassword(true, true, true, true, 16)?.length, 16);
      expect(
          Security.generatePassword(false, true, false, true, 12)?.length, 12);
      expect(
          Security.generatePassword(true, true, false, false, 25)?.length, 25);
      expect(
          Security.generatePasswordwithPrefix(
                  Security.generatePassword(true, true, false, false, 25)!,
                  "xert",
                  "_")
              ?.length,
          30);
    });
  });
}


lib/src/dart_encrypter_base.dart
import 'dart:math';
import 'package:encrypt/encrypt.dart';
import 'package:logbox_color/extensions.dart';
import 'package:logbox_color/logbox_color.dart';

///Security class.

class Security {
  ///* Info: Generate password.
  ///* Params: [isNumberActive] determines access code will include number 0-9 or not, [isSpecialCharacherActive] enables / disables special character, [isUpperCaseActive] enables / disables uppercase letters, [isLowerCaseActive] enables / disables lowercase letters, [length] is access code length.
  ///* Returns: String?
  ///* Notes:
  static String? generatePassword(
      bool isNumberActive,
      bool isSpecialCharacherActive,
      bool isUpperCaseActive,
      bool isLowerCaseActive,
      int length) {
    try {
//ASCII bounds.
      int minUpperLetter = 65;
      int maxUpperLetter = 90;

      int minLowerLetter = 97;
      int maxLowerLetter = 122;

      var specialChars = ["!", "#", "\$", "%", "&", "@", "*", "^"];

      StringBuffer sb = StringBuffer(); // String builder.
      Random rng = Random(); // Random number generator

      for (int i = 0; i < length; i++) {
        int choice;

        choice = rng.nextInt(
            4); // Four status available: * Uppercase letter, Lowerccase letter, number, special character

        if (choice == 0) // If 0 comes. lower a-z.
        {
          if (isLowerCaseActive) {
            int rangeLower = minLowerLetter +
                rng.nextInt((maxLowerLetter - minLowerLetter) + 1);
            sb.writeCharCode(rangeLower);
          } else {
            if (isUpperCaseActive) {
              int rangeUpper = minUpperLetter +
                  rng.nextInt((maxUpperLetter - minUpperLetter) + 1);
              sb.writeCharCode(rangeUpper);
              continue;
            }

            if (isNumberActive) {
              sb.write(rng.nextInt(10));
              continue;
            }

            if (isSpecialCharacherActive) {
              var selectID = rng.nextInt(specialChars.length);
              sb.write(specialChars[selectID]);
              continue;
            }
          }
        } else if (choice == 1) // 1 comes, upper A-Z.
        {
          if (isUpperCaseActive) {
            int rangeUpper = minUpperLetter +
                rng.nextInt((maxUpperLetter - minUpperLetter) + 1);
            sb.writeCharCode(rangeUpper);
          } else {
            if (isLowerCaseActive) {
              int rangeLower = minLowerLetter +
                  rng.nextInt((maxLowerLetter - minLowerLetter) + 1);
              sb.writeCharCode(rangeLower);
              continue;
            }
            if (isNumberActive) {
              sb.write(rng.nextInt(10));
              continue;
            }

            if (isSpecialCharacherActive) {
              var selectID = rng.nextInt(specialChars.length);
              sb.write(specialChars[selectID]);
              continue;
            }
          }
        } else if (choice == 2) // 2 comes. a number 0-9.
        {
          if (isNumberActive) {
            sb.write(rng.nextInt(10));
          } else {
            if (isLowerCaseActive) {
              int rangeLower = minLowerLetter +
                  rng.nextInt((maxLowerLetter - minLowerLetter) + 1);
              sb.writeCharCode(rangeLower);
              continue;
            }
            if (isSpecialCharacherActive) {
              var selectID = rng.nextInt(specialChars.length);
              sb.write(specialChars[selectID]);
              continue;
            }
            if (isUpperCaseActive) {
              int rangeUpper = minUpperLetter +
                  rng.nextInt((maxUpperLetter - minUpperLetter) + 1);
              sb.writeCharCode(rangeUpper);
              continue;
            }
          }
        } else // 3 comes. special characters.
        {
          if (isSpecialCharacherActive) {
            var selectID = rng.nextInt(specialChars.length);
            sb.write(specialChars[selectID]);
          } else {
            if (isUpperCaseActive) {
              int rangeUpper = minUpperLetter +
                  rng.nextInt((maxUpperLetter - minUpperLetter) + 1);
              sb.writeCharCode(rangeUpper);
              continue;
            }
            if (isLowerCaseActive) {
              int rangeLower = minLowerLetter +
                  rng.nextInt((maxLowerLetter - minLowerLetter) + 1);
              sb.writeCharCode(rangeLower);
              continue;
            }
            if (isNumberActive) {
              sb.write(rng.nextInt(10));
              continue;
            }
          }
        }
      }

      return sb.toString();
    } catch (e) {
      return null;
    }
  }

  ///* Info: Generate password with prefix. return password like "XXXXy'generatedpassword'". XXXX is prefix, y is delimiter.
  ///* Params:[password] is password,  [prefix] is prefix, [prefixDelimiter] is delimiter for prefix;
  ///* Returns: String?
  ///* Notes:
  static String? generatePasswordwithPrefix(
      String password, String prefix, String prefixDelimiter) {
    try {
      return "$prefix$prefixDelimiter$password";
    } catch (e) {
      printLog(e.toString(), LogLevel.error);
      return null;
    }
  }
}

/// Encryption string extension.
extension Encryption on String {
  ///* Info: Encyrpting the data.
  ///* Params:  [_key] is is Key String, [iv] is IV String.
  ///* Returns: String?
  ///* Notes: Key length must be == 32 , IV length must be == 16.
  String? encryptMyData(String _key, String iv) {
    try {
      if (_key.length == 32 && iv.length == 16) {
        final e = Encrypter(
            AES(Key.fromUtf8(_key), mode: AESMode.ctr, padding: null));
        final encryptedData = e.encrypt(this, iv: IV.fromUtf8(iv));
        String urlEncData = Uri.encodeComponent(encryptedData.base64);
        return urlEncData;
      } else {
        throw Exception("Length error. Check Key and IV Length.");
      }
    } catch (e) {
      printLog(e.toString(), LogLevel.error);
      return null;
    }
  }

  ///* Info: Decrypting the data.
  ///* Params:  [_key] is is Key String, [iv] is IV String.
  ///* Returns: String?
  ///*  Notes: Key length must be == 32 , IV length must be == 16.
  String? decryptMyData(String _key, String iv) {
    try {
      if (_key.length == 32 && iv.length == 16) {
        String urlDec = Uri.decodeComponent(this);
        final e = Encrypter(
            AES(Key.fromUtf8(_key), mode: AESMode.ctr, padding: null));
        final decryptedData =
            e.decrypt(Encrypted.fromBase64(urlDec), iv: IV.fromUtf8(iv));

        return decryptedData;
      } else {
        throw Exception("Length error. Check Key and IV Length.");
      }
    } catch (e) {
      printLog(e.toString(), LogLevel.error);
      return null;
    }
  }
}


example/example.dart
import 'package:dart_encrypter/dart_encrypter.dart';
import 'package:logbox_color/extensions.dart';
import 'package:logbox_color/logbox_color.dart';

void main() {
  final _key = Security.generatePassword(true, true, true, true, 32); //32 chars
  final iv = Security.generatePassword(true, true, true, true, 16); // 16 chars.

  printLog("Key: $_key", LogLevel.verbose); // Print key.

  printLog("IV: $iv", LogLevel.verbose); // Print IV.

  var item = "Hello my dear !".encryptMyData(_key!, iv!); //Example text.

  if (item != null) {
    // If item is not null.
    printLog("Encrypted text: $item", LogLevel.info); // Print encrypted text.
    var deItem = item.decryptMyData(_key, iv); // decrypt text.
    printLog("Decrypted text: $deItem", LogLevel.info); // Print decrypted text.
  }

  printLog(
      "Password 1: ${Security.generatePassword(true, true, true, true, 16)}",
      LogLevel.debug); // Length 16 password.
  printLog(
      "Password 2: ${Security.generatePassword(false, true, false, true, 12)}",
      LogLevel.debug); // Length 12 password.
  printLog(
      "Password 3: ${Security.generatePassword(true, true, false, false, 25)}",
      LogLevel.debug); // Length 25 password.

  printLog(
      "Password 4: ${Security.generatePassword(false, false, true, true, 12)}",
      LogLevel.debug); // Length 12 password.

  printLog(
      "Password 5: ${Security.generatePassword(true, false, false, false, 12)}",
      LogLevel.debug); // Length 12 password.

  printLog(
      "Password 6: ${Security.generatePassword(false, true, false, false, 12)}",
      LogLevel.debug); // Length 12 password.

  printLog(
      "Password with prefix: ${Security.generatePasswordwithPrefix(Security.generatePassword(true, true, false, false, 25)!, "xert", "_")}",
      LogLevel.debug); // Length 25 password with prefix.
}
